### PR TITLE
MAST in plugins, prevent empty/nameless archives, prevent hiding blacklisted files

### PR DESCRIPTION
**Pull request by Liderate**
- Change dummy plugins for Fallout 3 and New Vegas to include Fallout3.esm or FalloutNV.esm as master. This is required for BSAs that are trying to overwrite files from the vanilla BSAs. Archives are allowed to overwrite vanilla files with the .override file functionality from Command Extender or JIP LN NVSE.

- Prevent creating an archive when naming was cancelled, or name was empty. Some games do allow loading a nameless archive, but none I tested allow nameless plugins.

- Only enable the plugin for supported games.

- Get correct archive extension in a couple places.

- Prevent hiding files that have blacklisted extensions.

### CODE REVIEW

**Comment by Al12rs** [comment]
Hi, thanks for the contribution! 

Could you give a scenario for when the Blacklist for hiding would be necessary/useful?
Honestly I'm not terribly familiar with the hide loose files feature and when that would be used either, so I wanted to get some context.

Are you on the MO2 discord server? Don't know if you have a different handle there or simply aren't in it.

File: src/HideLooseAssetService.cpp [hunk 1]
```diff
@@ -19,12 +22,69 @@ namespace BsaPacker
 			return false;
 		}
 
+		// Get QStringList of blacklisted extensions and remove the . so we can match with QFileInfo::suffix later
```
Comment (Al12rs) [inline src/HideLooseAssetService.cpp]: Add comment:
`Only hide loose files that were not excluded when creating the Archive (blacklisted), so might still be needed by the mod`

File: src/HideLooseAssetService.cpp [hunk 2]
```diff
@@ -19,12 +22,69 @@ namespace BsaPacker
...
+				if (!fileDir.exists() && !fileDir.mkpath(fileDir.absolutePath())) {
+					qWarning() << "Failed to create path " << fileDir.absolutePath();
+					continue;
+				}
+
+				QString hiddenPath(blacklistedPath);
+				hiddenPath.insert(absPathLength, s_HiddenExt);
+				if (!fileDir.rename(hiddenPath, blacklistedPath)) {
+					qWarning() << "Failed to rename " << hiddenPath << " to " << blacklistedPath;
+				}
```
Comment (Al12rs) [inline src/HideLooseAssetService.cpp]: To simplify all this code I would suggest:
- Iterating over all files and saving a list blacklisted files found
- Hiding all folders
- Move back the blacklisted files

You can group the list of blacklisted files by first folder if it makes things easier.

**Review by Al12rs — CHANGES_REQUESTED** [review]

File: src/HideLooseAssetService.cpp (hunk 2 above)
Comment (Liderate) [reply src/HideLooseAssetService.cpp]: Thank you for the feedback. I have simplified it to:
- Modify blacklistExtensions to use as a filter by prepending a *
- Hide all folders
- Iterate through and move back matching files

File: src/HideLooseAssetService.cpp [hunk 3]
```diff
@@ -19,13 +23,35 @@ namespace BsaPacker
...
+		// Restore files with blacklisted extension to their original directories
+		QDir renamerDir(absModDir);
+		QDirIterator iterator(absModDir, blacklistExtensions, QDir::Files, QDirIterator::Subdirectories);
+		while (iterator.hasNext()) {
+			const QFileInfo& fileInfo(iterator.nextFileInfo());
+			if (!renamerDir.mkpath(fileInfo.absolutePath().replace(s_HiddenExt, ""))) {
+				qWarning() << "Failed to make directory " << fileInfo.absolutePath().replace(s_HiddenExt, "");
+				continue;
 			}
+			renamerDir.rename(fileInfo.absoluteFilePath(), fileInfo.absoluteFilePath().replace(s_HiddenExt, ""));
```
Comment (Al12rs) [inline src/HideLooseAssetService.cpp]: These `replace` are stripping `.mohidden` from the entire path, so potentially even from later down the tree, not just the top level.

If you have a blacklisted file that ends with `.mohidden`, then you are not only restoring it to the original unhidden folder but also removing the `.mohidden` part at the end, which is likely to be unwanted behaviour. 

You should only strip `.mohidden` from the first subdir name.

To make this simpler I think you could iterate over the hidden subdirs and find all blacklisted children inside each hidden subdir, and rename those, this way you have already the paths split at the right places:
- `parentDir / hiddenDir / blacklistedFilePath`
Which you change to :
- `parentDir / unhiddenDir / blacklistedFilePath`

Should also avoid iterating over the now unhidden blacklisted files that you already renamed.

**Review by Al12rs — CHANGES_REQUESTED** [review]
Looks much cleaner, thanks, there is still a potential issue

File: src/HideLooseAssetService.cpp (hunk 3 above)
Comment (Liderate) [reply src/HideLooseAssetService.cpp]: Good points. I have moved the iterator to be per hidden subdirectory. I also adjusted the iterator loop to address your comments. Let me know if you see any other issues (or new ones).

I also removed the gameDependency requirement for now. I overlooked some games that should be supported. I need to do more research about the games MO2 and the BSA Packer support before readding it. Using the tool on an unsupported game is handled gracefully, so it should be fine being visible in those instances.

File: src/HideLooseAssetService.cpp [hunk 4]
```diff
@@ -19,12 +22,35 @@ namespace BsaPacker
...
+				continue;
+			}
+
+			// Restore files with blacklisted extension to their original directories
+			QDir hiddenDir(originalDir.absolutePath() + s_HiddenExt);
+			QDirIterator iterator(hiddenDir.absolutePath(), blacklistExtensions, QDir::Files, QDirIterator::Subdirectories);
+			while (iterator.hasNext()) {
+				QString hiddenFilePath = iterator.next();
+				QString originalFilePath = originalDir.absoluteFilePath(hiddenDir.relativeFilePath(hiddenFilePath));
+				originalDir.mkpath(originalFilePath.left(originalFilePath.lastIndexOf("/")));
```
Comment (Al12rs) [inline src/HideLooseAssetService.cpp:50]: Ideally we would always use QFileInfo and QDir to do path operations, but the original code is guilty of this too.

**Review by Al12rs — APPROVED** [review]
Thanks for implementing the feedback!

**Pull request merged**